File: dotnet/Models/Transaction.cs
using System.ComponentModel.DataAnnotations;
using System.ComponentModel.DataAnnotations.Schema;

namespace InjectionInspection.Models
{
    /// <summary>
    /// Transaction entity with vulnerable search methods for security training
    /// Replicates the Python SQLAlchemy Transaction model exactly
    /// </summary>
    public class Transaction
    {
        [Key]
        public int Id { get; set; }

        [Required]
        [ForeignKey("User")]
        public int UserId { get; set; }

        [Required]
        [StringLength(20)]
        public string TransactionType { get; set; } = string.Empty; // 'credit' or 'debit'

        [Required]
        [Column(TypeName = "decimal(12,2)")]
        public decimal Amount { get; set; }

        [Required]
        [StringLength(100)]
        public string Company { get; set; } = string.Empty;

        public string? Description { get; set; }

        [Required]
        public DateTime Date { get; set; }

        [Required]
        [StringLength(50)]
        public string ReferenceNumber { get; set; } = string.Empty;

        [Required]
        [Column(TypeName = "decimal(12,2)")]
        public decimal BalanceAfter { get; set; }

        [StringLength(30)]
        public string? Category { get; set; }

        // User note with template injection vulnerability
        public string? Note { get; set; }

        // Navigation property
        public virtual User User { get; set; } = null!;

        // Helper methods matching Python implementation
        public bool IsCredit()
        {
            return TransactionType.Equals("credit", StringComparison.OrdinalIgnoreCase);
        }

        public bool IsDebit()
        {
            return TransactionType.Equals("debit", StringComparison.OrdinalIgnoreCase);
        }

        public string GetFormattedAmount()
        {
            var prefix = IsCredit() ? "+" : "-";
            return $"{prefix}${Amount:F2}";
        }

        public string GetFormattedDate()
        {
            return Date.ToString("MMM dd, yyyy");
        }

        public string GetFormattedDateTime()
        {
            return Date.ToString("MMMM dd, yyyy 'at' h:mm tt");
        }

        public static decimal GetTotalVolume(IEnumerable<Transaction> transactions)
        {
            return transactions.Sum(t => t.Amount);
        }

        public static decimal GetTotalCredits(IEnumerable<Transaction> transactions)
        {
            return transactions
                .Where(t => t.IsCredit())
                .Sum(t => t.Amount);
        }

        public static decimal GetTotalDebits(IEnumerable<Transaction> transactions)
        {
            return transactions
                .Where(t => t.IsDebit())
                .Sum(t => t.Amount);
        }

        public static List<Transaction> GetRecentActivity(IEnumerable<Transaction> transactions, int days = 30)
        {
            var cutoffDate = DateTime.UtcNow.AddDays(-days);
            return transactions
                .Where(t => t.Date >= cutoffDate)
                .OrderByDescending(t => t.Date)
                .ToList();
        }

        /// <summary>
        /// VULNERABLE: Generate reference number with predictable pattern
        /// Allows enumeration of transaction IDs
        /// </summary>
        public static string GenerateReferenceNumber(int transactionId)
        {
            // VULNERABILITY: Predictable reference numbers
            var timestamp = DateTime.UtcNow.ToString("yyyyMMdd");
            return $"TXN{timestamp}{transactionId:D6}";
        }

        /// <summary>
        /// SECURE: Generate cryptographically secure reference number (commented for training)
        /// </summary>
        // public static string GenerateSecureReferenceNumber()
        // {
        //     using var rng = RandomNumberGenerator.Create();
        //     byte[] bytes = new byte[16];
        //     rng.GetBytes(bytes);
        //     return Convert.ToHexString(bytes);
        // }
    }
}

File: nodejs/src/models/Transaction.ts
import { Entity, PrimaryGeneratedColumn, Column, ManyToOne, CreateDateColumn, Index, JoinColumn } from 'typeorm';
import { User } from './User';
import { AppDataSource } from '../database';

@Entity('transactions')
@Index(['userId'])
@Index(['date'])
export class Transaction {
    @PrimaryGeneratedColumn()
    id: number;

    @Column({ name: 'user_id', type: 'int' })
    userId: number;

    @Column({ name: 'transaction_type', type: 'varchar', length: 20 })
    transactionType: string; // 'debit' or 'credit'

    @Column({ type: 'decimal', precision: 12, scale: 2 })
    amount: number;

    @Column({ type: 'varchar', length: 100 })
    company: string;

    @Column({ type: 'text', nullable: true })
    description: string;

    @Column({ type: 'timestamp' })
    date: Date;

    @Column({ name: 'reference_number', unique: true, type: 'varchar', length: 50 })
    referenceNumber: string;

    @Column({ name: 'balance_after', type: 'decimal', precision: 12, scale: 2 })
    balanceAfter: number;

    @Column({ type: 'varchar', length: 30, nullable: true })
    category: string;

    @Column({ type: 'text', nullable: true })
    note: string;

    // Relationships
    @ManyToOne(() => User, user => user.transactions)
    @JoinColumn({ name: 'user_id' })
    user: User;

    // Instance methods
    formatAmount(): string {
        return `$${this.amount.toLocaleString('en-US', { minimumFractionDigits: 2, maximumFractionDigits: 2 })}`;
    }

    isDebit(): boolean {
        return this.transactionType === 'debit';
    }

    isCredit(): boolean {
        return this.transactionType === 'credit';
    }

    // Static methods
    static async getTotalVolume(): Promise<number> {
        const transactionRepo = AppDataSource.getRepository(Transaction);
        const result = await transactionRepo
            .createQueryBuilder('transaction')
            .select('SUM(transaction.amount)', 'total')
            .getRawOne();
        
        return parseFloat(result?.total || '0');
    }

    static async getTransactionCount(): Promise<number> {
        const transactionRepo = AppDataSource.getRepository(Transaction);
        return await transactionRepo.count();
    }

    static async getMonthlyVolume(): Promise<number> {
        const currentMonth = new Date();
        currentMonth.setDate(1);
        currentMonth.setHours(0, 0, 0, 0);
        
        const transactionRepo = AppDataSource.getRepository(Transaction);
        const result = await transactionRepo
            .createQueryBuilder('transaction')
            .select('SUM(transaction.amount)', 'total')
            .where('transaction.date >= :currentMonth', { currentMonth })
            .getRawOne();
        
        return parseFloat(result?.total || '0');
    }
}